### PR TITLE
handle websocket closed correctly

### LINKED ISSUE #3546: WebSockets do not fire 'close' event if the connection failed to be established
Consider the following code:
```js
const webSocket = new undici.WebSocket("wss://invalid-domain.example.com/");
webSocket.onopen = () => { console.log("open"); };
webSocket.onclose = () => { console.log("close"); };
webSocket.onerror = () => { console.log("error"); };
```
It outputs:
```
error
```

However, a standard-compliant WebSocket implementation would output:
```
error
close
```

This is because [establish a WebSocket connection](https://websockets.spec.whatwg.org/#concept-websocket-establish) algorithm would invoke [fail the WebSocket Connection](https://datatracker.ietf.org/doc/html/rfc6455#section-7.1.7) algorithm in step 11.1, which would [Close the WebSocket Connection](https://datatracker.ietf.org/doc/html/rfc6455#section-7.1.1). The WebSocket standard states:

> When [the WebSocket connection is closed](https://datatracker.ietf.org/doc/html/rfc6455#section-7.1.4), possibly cleanly, the user agent must [queue a task](https://html.spec.whatwg.org/multipage/webappapis.html#queue-a-task) to run the following substeps:
> ...
> 3. [Fire an event](https://dom.spec.whatwg.org/#concept-event-fire) named close at the [WebSocket](https://websockets.spec.whatwg.org/#websocket) object.

This bug is reproducible on commit [4b8958a](https://github.com/nodejs/undici/commit/4b8958a228aae55366dfba1dc70f03f08bc2d48d).

### PR DESCRIPTION
**Pull request by KhafraDev**
Fixes #3546 
Closes #3548

### CODE REVIEW

**Comment by KhafraDev** [comment]
I kinda forgot to run the autobahn tests locally, hopefully no issues

**Review by Uzlopak — APPROVED** [review]
LGTM

**Comment by eXhumer** [comment]
Related issue needs to be re-opened as I don't the issue is fixed. Infact, the test was modified which doesn't check the close event firing.

File: test/websocket/issue-3546.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+'use strict'
+
+const { test } = require('node:test')
+const { WebSocket } = require('../..')
+const { tspl } = require('@matteo.collina/tspl')
+
+test('first error than close event is fired on failed connection', async (t) => {
+  const { completed, strictEqual } = tspl(t, { plan: 2 })
```
Comment (eXhumer) [inline test/websocket/issue-3546.js:8]: `{ plan: 2 }` skips close check

**Comment by eXhumer** [comment]
Test output with fixed websocket test.

```
➜  undici git:(main) yarn test:websocket
yarn run v1.22.22
$ borp -p "test/websocket/*.js"
▶ WebSocketInit
  ✔ WebSocketInit as 2nd param (37.936791ms)
▶ WebSocketInit (38.6495ms)
▶ isValidSubprotocol
  ✔ empty string returns false (2.16575ms)
  ✔ simple valid value returns false (1.865417ms)
  ✔ empty string returns false (1.099334ms)
  ✔ value with "(),/:;<=>?@[\]{} returns false (0.762792ms)
▶ isValidSubprotocol (6.949709ms)
✔ Sending >= 2^16 bytes (43.259375ms)
✔ Sending >= 126, < 2^16 bytes (6.401209ms)
✔ Sending < 126 bytes (4.908042ms)
✔ Sending data after close (8.025542ms)
✔ Sending data before connected (2.06475ms)
▶ Sending data to a server
  ✔ Send with string (11.411375ms)
  ✔ Send with ArrayBuffer (8.749917ms)
  ✔ Send with Blob (13.136917ms)
  ✔ Cannot send with SharedArrayBuffer (12.56ms)
▶ Sending data to a server (46.437ms)
✔ check cloned (44.237375ms)
✔ Receiving a frame with a payload length > 2^31-1 bytes (39.807084ms)
✔ Receiving an ArrayBuffer (7.472417ms)
✔ Receives ping and parses body (40.736ms)
✔ Receives pong and parses body (6.461583ms)
✔ WebSocket connecting to server that isn't a Websocket server (39.134292ms)
✔ Open event is emitted (6.276667ms)
✔ Multiple protocols are joined by a comma (5.4405ms)
✔ Server doesn't send Sec-WebSocket-Protocol header when protocols are used (10.549625ms)
✔ Server sends invalid Upgrade header (4.053292ms)
✔ Server sends invalid Connection header (5.831458ms)
✔ Server sends invalid Sec-WebSocket-Accept header (12.565042ms)
✔ Server sends invalid Sec-WebSocket-Extensions header (7.32725ms)
✔ Server sends invalid Sec-WebSocket-Extensions header (15.754334ms)
✔ test/parallel/test-messageevent-brandcheck.js (2.147916ms)
✔ test/parallel/test-worker-message-port.js (3.207208ms)
✔ bug in node core (0.144083ms)
✖ first error than close event is fired on failed connection (8.1675ms)
  'Promise resolution is still pending but the event loop has already resolved'

✔ readyState is set on fail (5.174041ms)
✔ Receiving frame with payload length 0 works (25.618208ms)
✔ Fragmented frame with a ping frame in the first of it (24.409375ms)
✔ The server must reply with at least one subprotocol the client sends (36.308958ms)
✔ The connection fails when the client sends subprotocols that the server does not responc with (9.595459ms)
✔ Close without receiving code does not send an invalid payload (3029.1465ms)
✔ Don not use pooled buffer in mask pool (1.11575ms)
✔ Writing 16-bit frame length value at correct offset when buffer has a non-zero byteOffset (0.320625ms)
✔ Fragmented frame with a ping frame in the middle of it (34.053959ms)
✔ MessageEvent (1.460375ms)
✔ CloseEvent (0.304583ms)
✔ ErrorEvent (0.1585ms)
▶ Event handlers
  ▶ onopen
    ✔ should be null initially (0.119375ms)
    ✔ should not allow non-function assignments (0.071ms)
    ✔ should allow function assignments (0.122375ms)
  ▶ onopen (0.424584ms)
  ▶ onerror
    ✔ should be null initially (0.156708ms)
    ✔ should not allow non-function assignments (0.193292ms)
    ✔ should allow function assignments (0.250833ms)
  ▶ onerror (0.908125ms)
  ▶ onclose
    ✔ should be null initially (0.113292ms)
    ✔ should not allow non-function assignments (0.05725ms)
    ✔ should allow function assignments (0.04375ms)
  ▶ onclose (0.272625ms)
  ▶ onmessage
    ✔ should be null initially (0.062375ms)
    ✔ should not allow non-function assignments (0.043625ms)
    ✔ should allow function assignments (0.034375ms)
  ▶ onmessage (0.231541ms)
▶ Event handlers (4.203041ms)
▶ CloseEvent WPTs ported
  ✔ initCloseEvent (0.073625ms)
  ✔ CloseEvent constructor (0.096833ms)
▶ CloseEvent WPTs ported (0.225792ms)
▶ ErrorEvent WPTs ported
  ✔ Synthetic ErrorEvent (0.173583ms)
  ✔ webidl (0.062625ms)
  ✔ initErrorEvent (0.034625ms)
▶ ErrorEvent WPTs ported (0.318875ms)
▶ diagnostics channel
  ✔ undici:websocket:open (36.164959ms)
  ✔ undici:websocket:close (5.774417ms)
▶ diagnostics channel (43.490792ms)
✔ Setting custom headers (3.430959ms)
✔ Receiving multiple continuation frames works as expected (28.503459ms)
✔ Constructor (1.908542ms)
▶ Close
  ✔ Close with code (23.061959ms)
  ✔ Close with code and reason (3.295125ms)
  ✔ Close with invalid code (3.016375ms)
  ✔ Close with invalid reason (4.419167ms)
  ✔ Close with no code or reason (3.371834ms)
  ✔ Close with a 3000 status code (2.238459ms)
  ✔ calling close twice will only trigger the close event once (2.692875ms)
▶ Close (43.034666ms)
✔ Receiving a close frame with invalid utf-8 (25.635125ms)
✔ Client fails the connection if receiving a masked frame (24.697625ms)
✔ Client fails the connection if receiving a masked frame (23.566167ms)
ℹ tests 76
ℹ suites 12
ℹ pass 75
ℹ fail 0
ℹ cancelled 1
ℹ skipped 0
ℹ todo 0
ℹ duration_ms 3528.956583

✖ failing tests:

test at test/websocket/issue-3546.js:7:1
✖ first error than close event is fired on failed connection (8.1675ms)
  'Promise resolution is still pending but the event loop has already resolved'
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
➜  undici git:(main) ✗
```

**Pull request merged**